File: product-service/src/controller/product.controller.ts
/**
 * Author: Le Anh Duc
 * Email: ducvui2003@gmail.com
 * Phone number: +84 965809127
 * Created at: 28/8/24 - 4:59 pm
 * User: ducvui2003
 **/
import {NextFunction, Request, Response} from "express";
import ProductService from "../service/product.service";
import {ResponseLocals} from "../type/response.type";
import {DiscountInfoDocument, ProductDocument} from "../document/product.document";
import {ProductModel} from "../model/product.model";
import ApiPagingType from "../type/apiPaging.type";

const getById = async (req: Request, res: Response<any, ResponseLocals<ProductModel>>, next: NextFunction) => {
    try {
        res.locals.data = await ProductService.getById(req.params.id)
        next()
    } catch (e) {
        next(e)
    }
}

const getAll = async (req: Request<{}, {}, {}, {
    page: string
}, {}>, res: Response<any, ResponseLocals<ApiPagingType<ProductModel>>>, next: NextFunction) => {
    const page = req.query.page ? Number.parseInt(req.query.page) : 1
    res.locals.data = await ProductService.getAll(page)
    next()
}

const create = async (req: Request<any, any, ProductDocument>, res: Response<any, ResponseLocals<ProductModel>>, next: NextFunction) => {
    res.locals.data = await ProductService.create(req.body)
    next();
}

const removeDiscount = async (req: Request<{
    id: string
}>, res: Response<any, ResponseLocals<ProductModel>>, next: NextFunction) => {
    try {
        res.locals.data = await ProductService.removeDiscount(req.params.id)
        next();
    } catch (e) {
        next(e)
    }
}

const setDiscount = async (req: Request<{ id: string }, any, DiscountInfoDocument>, res: Response<any, ResponseLocals<ProductModel>>, next: NextFunction) => {
    try {
        res.locals.data = await ProductService.setDiscount(req.params.id, req.body)
        next();
    } catch (e) {
        next(e)
    }
}

const updateUrlImage = async (req: Request<{ id: string }, {}, {
    url: string
}>, _: Response<any, ResponseLocals<boolean>>, next: NextFunction) => {
    try {
        await ProductService.updateUrlImage(req.params.id, req.body.url)
        next();
    } catch (e) {
        next(e)
    }
}

const getByCategory = async (req: Request<{ id: string }, {}, {}, {
    page: string
}, {}>, res: Response<any, ResponseLocals<ApiPagingType<ProductModel>>>, next: NextFunction) => {
    const page = req.query.page ? Number.parseInt(req.query.page) : 1
    res.locals.data = await ProductService.getByCategory(req.params.id, page)
    next()
}


export default {getById, getAll, create, removeDiscount, setDiscount, updateUrlImage, getByCategory}

File: src/main/java/com/spring/delivery/controller/ProductController.java
package com.spring.delivery.controller;

import com.spring.delivery.domain.ApiPaging;
import com.spring.delivery.domain.request.product.*;
import com.spring.delivery.domain.response.product.CardProductDTO;
import com.spring.delivery.domain.response.product.ProductDTO;
import com.spring.delivery.domain.response.product.ProductOptionDTO;
import com.spring.delivery.service.business.product.IProductOptionService;
import com.spring.delivery.service.business.product.IProductService;
import com.spring.delivery.service.business.user.IUserProductFavoriteService;
import com.spring.delivery.util.anotation.ApiMessage;
import jakarta.validation.Valid;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/product")
@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
@RequiredArgsConstructor
public class ProductController {
    IProductService productService;
    IUserProductFavoriteService userProductFavoriteService;

    @GetMapping("/{id}")
    @ApiMessage("Get product by id")
    public ResponseEntity<ProductDTO> getById(@PathVariable("id") String id) {
        return ResponseEntity.ok(productService.findById(id));
    }

    @GetMapping
    @ApiMessage("Get all product per page")
    public ResponseEntity<ApiPaging<CardProductDTO>> getAll(@RequestParam(required = false, name = "page", defaultValue = "1") int page) {
        return ResponseEntity.ok(productService.findAll(page));
    }

    @GetMapping("/category/{id}")
    @ApiMessage("Get product by id category")
    public ResponseEntity<ApiPaging<CardProductDTO>> getByCategory(@PathVariable("id") String id, @RequestParam(required = false, name = "page", defaultValue = "1") int page) {
        return ResponseEntity.ok(productService.findAllByCategoryId(id, page));
    }

    @PostMapping(consumes = "application/json;charset=UTF-8")
    @PreAuthorize("hasRole('ADMIN')")
    @ApiMessage("Create product")
    public ResponseEntity<ProductDTO> createProduct(@Valid @RequestBody RequestProductCreated request) {
        return ResponseEntity.ok(productService.save(request));
    }

    @PutMapping(path = "/{id}", consumes = "application/json;charset=UTF-8")
    @PreAuthorize("hasRole('ADMIN')")
    @ApiMessage("Create product")
    public ResponseEntity<ProductDTO> updateProduct(@PathVariable("id") String id, @Valid @RequestBody RequestProductUpdated request) {
        return ResponseEntity.ok(productService.updateProduct(id, request));
    }

    @DeleteMapping("/discount/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @ApiMessage("Remove discount product")
    public ResponseEntity<ProductDTO> removeDiscount(@PathVariable("id") String id) {
        return ResponseEntity.ok(productService.removeDiscount(id));
    }

    @PutMapping("/discount/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @ApiMessage("Set discount for product")
    public ResponseEntity<ProductDTO> setDiscount(@PathVariable("id") String id, @Valid @RequestBody RequestDiscountCreated request) {
        return ResponseEntity.ok(productService.setDiscount(id, request));
    }

    @PutMapping("/image/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @ApiMessage("Update url image product")
    public ResponseEntity<ProductDTO> updateUrlImage(@PathVariable("id") String id, @Valid @RequestBody RequestUpdateImage request) {
        return ResponseEntity.ok(productService.updateUrlImage(id, request));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') AND hasAuthority('PRODUCT_DELETE')")
    @ApiMessage("Delete product by id")
    public ResponseEntity<ProductDTO> deleteProduct(@PathVariable("id") String id) {
        return ResponseEntity.ok(productService.deleteProduct(id));
    }

    @PostMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') AND hasAuthority('PRODUCT_DELETE')")
    @ApiMessage("Undelete product by id")
    public ResponseEntity<ProductDTO> unDeleteProduct(@PathVariable("id") String id) {
        return ResponseEntity.ok(productService.unDeleteProduct(id));
    }

    @GetMapping("/home-page")
    @ApiMessage("Find product for home page")
    public ResponseEntity<List<ProductDTO>> findProductForHomePage() {
        return ResponseEntity.ok(productService.findProductForHomePage());
    }

    @GetMapping("/search")
    @ApiMessage("Search product")
    public ResponseEntity<ApiPaging<CardProductDTO>> searchProduct(RequestSearchProduct request) {
        return ResponseEntity.ok(productService.searchProduct(request));
    }

    @PostMapping("/favorite/{id}")
    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    @ApiMessage("Favorite product success")
    public ResponseEntity<Void> favorite(@PathVariable("id") String id) {
        userProductFavoriteService.addProductFavorite(id);
        return ResponseEntity.ok(null);
    }

    @DeleteMapping("/favorite/{id}")
    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    @ApiMessage("Remove favorite product success")
    public ResponseEntity<Void> unFavorite(@PathVariable("id") String id) {
        userProductFavoriteService.removeProductFavorite(id);
        return ResponseEntity.ok(null);
    }

    @GetMapping("/favorite")
    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    @ApiMessage("Favorite product")
    public ResponseEntity<ApiPaging<CardProductDTO>> getFavorite(RequestSearchProduct request, @PageableDefault(sort = "id") Pageable pageable) {
        var result = productService.getFavorite(request, pageable);
        return ResponseEntity.ok().body(result);
    }
}
